lib/charms/opensearch/v0/opensearch_relation_provider.py

```diff
@@ -359,20 +364,27 @@ def update_certs(self, relation_id, ca_chain=None):
         """Update TLS certs passed into this relation.
 
         If ca_chain is not provided, it'll get the app-admin CA generated by the TLS charm.
+
+        Raises:
+            KeyError: if the CA chain can't be found in the secret.
         """
         if not self.unit.is_leader():
             # We're updating app databags in this function, so it can't be called on follower
             # units. This is not checked in `set_tls_ca`, in data-interfaces.
             return
         try:
             # If the ca_chain=None, then we try to load the CA from the app-admin secret.
-            _ch_chain = ca_chain or self.charm.secrets.get_object(
-                Scope.APP, CertType.APP_ADMIN.val
-            ).get("chain")
+            # If neither ca_chain nor the secret exists, we'll raise a KeyError.
+            _ch_chain = (
+                ca_chain
+                or self.charm.secrets.get_object(Scope.APP, CertType.APP_ADMIN.val)["chain"]
+            )
         except AttributeError:
             # cert doesn't exist - presumably we don't yet have a TLS relation.
             logger.warning("unable to get ca_chain")
             return
+        except TypeError:
```

Comment:
I'm confused, shouldn't this be a KeyError if the `["chain"]` doesn't exist (as per [Built-in Exceptions](https://docs.python.org/3/library/exceptions.html#KeyError))?

Reply:
Yes, and in this case, we let it be raised and we catch it on the calling methods.

Now, if you try instead `None["chain"]`, then you have a type (`None`) being wrongly used as another type (`dict`). That is a TypeError and it is also a possibility here if nothing is set.

To simplify, I just let KeyError be raised.

Reply:
In my opinion, I think it would be cleaner to raise a custom exception rather than raising the TypeError as a KeyError

Reply:
+1 with @skourta - we should not propagate a `KeyError`. There should be a TLS related exception raised / caught in the consumer side. 

Why are we also handling `AttributeError`? 

We can I believe simplify this part as:
```
admin_secrets = self.charm.secrets.get_object(Scope.APP, CertType.APP_ADMIN.val, {})
if not (_ch_chain := ca_chain or admin_secrets.get("chain")):
    raise TLSCaChainMissingError()
```

Reply:
@Mehdi-Bendriss on the Error, okay, I will add another error class here (in another PR and ref your comment).

About `AttributeError`, I think the reason there was because original code was:

```
... .get("chain")
```

Which may result in a `None.get("chain")`. I looked into the code underneath the secret calls but it handles None carefully, so there is no way I could see an `AttributeError` bubbling up. TL;DR it is safe to remove that 2nd exception.